### ISSUE TITLE
The hardware requirements

### ISSUE THREAD

**Issue by Jueqian** [PRBonn/3DiSS#3]
May I ask the minimum hardware requirements are for running training and inference?
Can RTX3090 24G run it well?

**Comment by nuneslu** [comment]
Hi! For inference, it should be fine. For training, you would need to reduce the batch size. Another option is to supervise only a subset of the output voxels. Currently, we compute the loss over all the output voxels, you can also supervise only for half of the voxels (since we trained with 48GB GPUs).

**Comment by Jueqian** [comment]
Hi Lucas, when I using 8*L20 (with 48GB) to train vae and diffusion model with default configuration provided in the repository, there were instances of OOM errors during both VAE training and diffusion training. The training process cannot complete a full epoch.

More specifically, during VAE training, GPU 0 runs into OOM immediately at the start. During diffusion training, GPU 0 uses about twice as much memory as the other GPUs and goes OOM at around 130 iterations.

Do you have such a situation?

**Comment by nuneslu** [comment]
Hi @Jueqian! I did have this problem at some point, but it did not happen again. A few hints to overcome it:

- MinkowskiEngine (ME) GPU usage fluctuates a lot, given that it uses sparse operations. One problem that I noticed is that pytorch does not handle it in a smart way, so in the very first iteration, using ME pytorch allocates more memory than it actually needs, then, from the second iteration on, it allocates the correct amount. So, for the crash on the first iteration, this might be the cause.
- Regarding the problem with GPU 0 allocating more memory, it is a problem with how lightning handles the distributed processing. Therefore, the more GPUs you use in parallel, the higher the overload of GPU 0. However, I recently have managed to run the code with a more recent version of torch and lightning (`torch==1.12.1+cu116` and `pytorch_lightning==2.1.0`), and it seems that from 2.0.0 on, lightning fixed this problem with distributed processing.

**TL;DR**

I suggest trying to reduce the number of GPUs used for training. Or using `torch==1.12.1+cu116` and `pytorch_lightning==2.1.0`, where the distributed processing better handles the GPU usage. I will also push the Dockerfile I used for running it with the more recent torch and lightning versions.

**Comment by Jueqian** [comment]
Thans for your detailed response! After switching the version to `torch==1.12.1+cu116` and  `pytorch_lightning==2.1.0` as you suggested, the GPU memory distribution became more balanced, and the diffusion model can now train properly.

For the first point you mentioned, I noticed that in your code the first iteration uses only a single batch (https://github.com/PRBonn/3DiSS/blob/46eced27941db043957eec8dc59fda62418bdf1f/diss/models/sem_vae.py#L153-L158), but it doesn’t seem to work. After I changed the batch size to 1, the training runs successfully. Indeed, after the first iteration, the GPU memory usage drops and becomes relatively stable.

I also noticed that training the VAE model takes quite a long time (with batch size = 1, training one epoch on six L20 GPUs takes about 2.5 hours. In your paper, you mentioned training for 50 epochs, which would be roughly 125 hours?). Since the training time is quite long, I’m temporarily using the VAE model you provided to train the diffusion model.

Sincerely thank you again and wish you all the best! 😄
